### PR TITLE
Mongodb DAO Update

### PR DESCRIPTION
**Pull request by PatrickFuhrmann-HTWBerlin**
## Main Objective:

The goal of this pull request is to prepare for adding new formats to the OAI PMH provider in addition to Panosc, dc and openaire. As a first step we would like to get the mongodb DAO and the entire package updated to the most recent version.
The pull request consists of various commits with the corresponding explanation.  

### We are running:

- npm --version : 10.5.1
- node --version: v22.0.0

- A) on 'npm install':
When doing a npm install, you will find WARNINGS on 'inflight' and 'glob'. They are hard to fix because they are loaded from gulp. But they are not so important as they are only used for the installation while using gulp. The actual services is not using them.
- B) I didn't touch the CI/ESS part. I don't know why they have the private directory.
- C) We only updated 'production/.env' to add the key 'COLLECTION_ID'. This is only important when not using the COLLECTION_ID='doi'. As a general remark:  When running docker, docker-compose or kubernetes, you will overwrite those values anyway, so the content is not that important.
- D) We are working on the workflow part now, as Carlo wants us to push a docker image to ghcr.io from within a .githup/workflow. 
- E) Works for us
   - 'npm ci'
   - 'npm run compile'
   - docker build -t oai-pmh-provider' .
- F) I checked the functionality by building the docker container by the Dockerfile and running it within scicat-live. The following REST calls produce the correct output:
     http://localhost:3000/
     http://localhost:3000/scicat/oai?verb=Identify
     http://localhost:3000/scicat/oai?verb=ListMetadataFormats
     http://localhost:3000/scicat/oai?verb=GetRecord&identifier=1&metadataPrefix=oai_dc
     http://localhost:3000/scicat/oai?verb=ListIdentifiers&metadataPrefix=oai_dc
     http://localhost:3000/scicat/oai?verb=ListRecords&metadataPrefix=oai_dc
 
The same is true for the '/panosc/oai' and '/openaire/oai' URL. I must admit, I didn't try to 'PUT' a record, because records are added through the backend-next/frontend system and PUT is not mentioned in the README.md.

We are happy to improve the pull request. We might not be 100% compliant with your coding rules.

-p.

### CODE REVIEW

File: src/providers/scicat-provider/dao/mongo-dao.ts [hunk 1]
```diff
@@ -13,6 +13,7 @@ export class MongoConnector {
   public db;
   public dbName: string;
   public collectionName: string;
+  public mongoDb: MongoClient ;
```
Comment (minottic) [inline src/providers/scicat-provider/dao/mongo-dao.ts:16]: minor comment: extra space before `;`

File: src/providers/scicat-provider/dao/mongo-dao.ts [hunk 2]
```diff
@@ -24,13 +25,20 @@ export class MongoConnector {
...
     this.collectionName = process.env.COLLECTION;
 
-    MongoClient.connect("mongodb://" + url, { useUnifiedTopology: true }, (err, client) => {
-      if (err) {
-        logger.error("failed to connect", err);
-        this.db = null;
-      }
-      this.db = client.db(this.dbName);
-    });
+     const  myUrl = "mongodb://"+url ;
```
Comment (minottic) [inline src/providers/scicat-provider/dao/mongo-dao.ts:28]: I would call it `mongoUrl`, I find it better to give variables names with a meaning

File: src/providers/scicat-provider/dao/mongo-dao.ts [hunk 3]
```diff
@@ -50,68 +58,43 @@ export class MongoConnector {
...
-      // need to add relevant date to projection
-      Publication.find(filter, { _id: 1 }).toArray(function(err, items) {
-        if (err) {
-          reject(err);
-        } else {
-          resolve(items);
-        }
-      });
-    });
+    return await Publication.find(filter /* , { _id: 1 } */).toArray() ;
```
Comment (minottic) [inline src/providers/scicat-provider/dao/mongo-dao.ts:79]: I think the `{_id: 1}` was done to only return the _id from the query and not the full document. Is this commented on purpose?

File: src/providers/scicat-provider/dao/mongo-dao.ts
Comment (minottic) [inline src/providers/scicat-provider/dao/mongo-dao.ts:1]: formatting is a little messy (some extra spaces and some missing spaces). not the end of the world though. 

Just a comment: conceptually this commit does 2 things:
1. comply with mongo 6 
2. refactor promises to async/await

While the 1 is required if willing to use mongo 6, the second is more of a cosmetic change, so it could live on its own or even be left for later. 

I am fine to keep it as is, it's just a "conceptual" comment

File: src/providers/scicat-provider/dao/mongo-dao.ts [hunk 4]
```diff
@@ -50,68 +58,43 @@ export class MongoConnector {
    * @param parameters
    * @returns {Promise<any>}
    */
-  public recordsQuery(parameters: any, filter: MongoClient.filter): Promise<any> {
+  public async recordsQuery(parameters: any, filter: any): Promise<any> {
```
Comment (minottic) [inline src/providers/scicat-provider/dao/mongo-dao.ts:61]: I think you can now import the filter from 
```
import { Filter } from 'mongodb';
```
so that we can keep type check

File: ENVIRONMENT_README.md
Comment (minottic) [inline ENVIRONMENT_README.md:1]: why don't you add this as a block in the main `README.md`, so that is rendered by github when users browse to it?

File: Dockerfile [hunk 5]
```diff
@@ -5,6 +5,8 @@ WORKDIR /home/node/app
 COPY package*.json /home/node/app/
 COPY . /home/node/app/
 
+RUN npm install -g npm@10.8.1
```
Comment (minottic) [inline Dockerfile:8]: why is this needed?

File: gulpfile.js [hunk 6]
```diff
@@ -47,8 +46,17 @@ gulp.task("copy-production", function() {
...
-gulp.task("dist-clean", function() {
-    return del(["dist/**/*", "dist/server/.env"]);
-});
 
 
+const rimraf = require('rimraf');
+
+gulp.task("dist-clean", function(done) {
+
+try {
```
[the remark this answers is not in the document]
Comment (PatrickFuhrmann-HTWBerlin) [reply gulpfile.js]: Is fixed.

File: gulpfile.js
[the remark this answers is not in the document]
Comment (PatrickFuhrmann-HTWBerlin) [reply gulpfile.js:1]: True, but I didn't get it work work with the old libaray.

File: src/providers/scicat-provider/dao/mongo-dao.ts (hunk 1 above)
Comment (PatrickFuhrmann-HTWBerlin) [reply src/providers/scicat-provider/dao/mongo-dao.ts]: Fixed

File: src/providers/scicat-provider/dao/mongo-dao.ts (hunk 2 above)
Comment (PatrickFuhrmann-HTWBerlin) [reply src/providers/scicat-provider/dao/mongo-dao.ts]: Fixed

File: src/providers/scicat-provider/dao/mongo-dao.ts (hunk 3 above)
Comment (PatrickFuhrmann-HTWBerlin) [reply src/providers/scicat-provider/dao/mongo-dao.ts]: Fixed. Just forgot to remove it.

File: src/providers/scicat-provider/dao/mongo-dao.ts
Comment (PatrickFuhrmann-HTWBerlin) [reply src/providers/scicat-provider/dao/mongo-dao.ts:1]: Split the async/await from the formatting fixes and the 'Filter' fix.

File: src/providers/scicat-provider/dao/mongo-dao.ts (hunk 4 above)
Comment (PatrickFuhrmann-HTWBerlin) [reply src/providers/scicat-provider/dao/mongo-dao.ts]: Fixed. [4485e2e](https://github.com/SciCatProject/oai-provider-service/pull/112/commits/4485e2e94707a02121986363c1282012b0164fea)

File: Dockerfile (hunk 5 above)
Comment (PatrickFuhrmann-HTWBerlin) [reply Dockerfile:8]: Otherwise it always complains about a missing upgrade.

File: ENVIRONMENT_README.md
Comment (PatrickFuhrmann-HTWBerlin) [reply ENVIRONMENT_README.md:1]: Fixed: [8e50d62](https://github.com/SciCatProject/oai-provider-service/pull/112/commits/8e50d628cf0108ac2218fb8167ee97ee84267132) and [658b721](https://github.com/SciCatProject/oai-provider-service/pull/112/commits/658b7212600ae3c66b06f126f720e0be57ef3698)

**Comment by PatrickFuhrmann-HTWBerlin** [comment]
If fixed the Tests.  The way the tests are done are a bit fragile. The response to the API calls are not checked against XML spec of XMl namespace but are a char to char comparison.
* I got the test to succeed in my repository but don't know if it works in the scicat one.
* I took the deploy.yml from BE as suggested but removed the ESS pipeline for now.

File: .github/workflows/deploy.yml [hunk 7]
```diff
@@ -0,0 +1,40 @@
+name: Deploy
+
+on:
+  push:
+    branches: [ "mongo-workflow-fix" ]
```
Comment (bpedersen2) [inline .github/workflows/deploy.yml:5]: OK for testing, but should get changed to master just before merge

File: README.md [hunk 8]
```diff
@@ -56,6 +57,69 @@ Add new records to your mongodb instance by HTTP PUT using the following route:
...
 
+## ENVIRONMENT Variables
+
+*System variables**
+ Key | Description | Default
+ --------:|-------------| --------
+DAPP_ID | No Idea | oai-pmh-service
+CONNECTOR | (don't change) | mongodb
+ADMIN_USER_EMAIL | E-Mail address of the admin user | none
+LOG_LEVEL | default/error/warning | none
```
Comment (bpedersen2) [inline README.md:68]: default should probably be `error`, but that can be done in an other change

File: README.md [hunk 9]
```diff
@@ -56,6 +57,69 @@ Add new records to your mongodb instance by HTTP PUT using the following route:
...
+**Mongo DB variables**
+ Key | Description | Default
+ --------:|-------------| --------
+CONNECTOR | (don't change) | mongodb
+DB_HOST | Database Hostname | none
+DB_PORT | Database Port | none
+DB_USER | Database Username | none
+DB_PASS | Database Password | none
+DB_URL |  [&lt;user&gt;:&lt;password&gt;@]&lt;host&gt;:&lt;port&gt;/&lt;dbName&gt;| none
+DATABASE | Publication Database | dacat-next
```
Comment (bpedersen2) [inline README.md:79]: maybe change to new scicat namings in a follow-up

File: package.json
[the remark this answers is not in the document]
Comment (bpedersen2) [reply package.json:1]: Agreed. Changes look OK here

File: test/mongo-dao.ts [hunk 10]
```diff
@@ -18,3 +18,13 @@ describe(`Mongo dao`, () => {
       {thumbnail: 0, doi: 0, publisher: 1, title: 1})
   })
 })
+
+describe('Waiting for db connection', function() {
+    it('should wait awhile before continuing', function(done) {
+        this.timeout(61000); // Optional: set timeout for this test to 61 seconds
+        setTimeout(function() {
+            // Your test logic here
```
Comment (bpedersen2) [inline test/mongo-dao.ts:26]: currently a no-op test?

**Review by bpedersen2 — COMMENTED** [review]
LGTM

File: src/providers/scicat-provider/dao/mongo-dao.ts [hunk 11]
```diff
@@ -2,7 +2,7 @@ import logger from "../../../server/logger";
 import { reject } from "bluebird";
 import { MongoClient } from "mongodb";
 import { getCollectionID } from "../../../server/env";
-
+import { Filter } from 'mongodb';
```
Comment (minottic) [inline src/providers/scicat-provider/dao/mongo-dao.ts:5]: should this be part of the import on top?

```
import { MongoClient, Filter } from "mongodb";
```

File: src/providers/scicat-provider/dao/mongo-dao.ts [hunk 12]
```diff
@@ -71,20 +73,20 @@ export class MongoConnector {
...
    * @returns {Promise<any>}
    */
-  public async identifiersQuery(parameters: any, filter: any): Promise<any> {
+  public async identifiersQuery(parameters: any, filter: Filter): Promise<any> {
     if (!this.db) {
       reject("no db connection");
     }
     let Publication = this.db.collection(this.collectionName);
-    return await Publication.find(filter /* , { _id: 1 } */).toArray() ;
+    return await Publication.find(filter).toArray() ;
```
Comment (minottic) [inline src/providers/scicat-provider/dao/mongo-dao.ts]: is the {_id : 1} removed on purpose?

**Review by minottic — COMMENTED** [review]
LGTM, thanks. 2 minor comments, in addition to Bjorn's.

Just a side note. In principle this PR should include tests for the modified functions (as they aren't there already). The existing tests don't really test the mongo functionality as they mock their behaviour. I am fine mergining without, anyway these changes are mostly using mongo native functionalities that we assume have been tested by mongo

File: test/mongo-dao.ts (hunk 10 above)
Comment (PatrickFuhrmann-HTWBerlin) [reply test/mongo-dao.ts:26]: True. But it creates a 10 second delay which might not be very elegant. But I had to make sure that the oai-pmh-service part had a chance to connect to the mongodb server. Otherwise all tests fail.

File: .github/workflows/deploy.yml (hunk 7 above)
Comment (PatrickFuhrmann-HTWBerlin) [reply .github/workflows/deploy.yml]: Fixed with [0dc0853](https://github.com/SciCatProject/oai-provider-service/pull/112/commits/0dc085392036af780768ba19704ca479c0e3dfaf)

File: README.md (hunk 8 above)
Comment (PatrickFuhrmann-HTWBerlin) [reply README.md]: Fixed with [95b512b](https://github.com/SciCatProject/oai-provider-service/pull/112/commits/95b512b51d755c65dad1ee0d12fbb0dfc953a9d8)

File: src/providers/scicat-provider/dao/mongo-dao.ts (hunk 12 above)
Comment (PatrickFuhrmann-HTWBerlin) [reply src/providers/scicat-provider/dao/mongo-dao.ts]: By mistake. Fixed with  738df60

File: README.md (hunk 9 above)
Comment (PatrickFuhrmann-HTWBerlin) [reply README.md:79]: Ok will do later.

**Comment by PatrickFuhrmann-HTWBerlin** [comment]
I had to fix two more problems:

- [c16411c](https://github.com/SciCatProject/oai-provider-service/pull/112/commits/c16411ca7388cfb82976948fff51a5636b42ff67): On githup (not on my mac M1 and not on my ARM reference) the node:18-alpine produced errors. This doesn't happen with the node:18-slim.
- 0c81791: Using the stricter and more typesafe Filter<> over Filter syntax to satisfy the gulp ts compiler.

File: .github/workflows/ci.yml [hunk 13]
```diff
@@ -1,6 +1,10 @@
 name: CI
 
-on: [pull_request]
+on:
+  push:
```
Comment (minottic) [inline .github/workflows/ci.yml:4]: maybe this ca be on PR only?

File: Dockerfile [hunk 14]
```diff
@@ -1,10 +1,12 @@
-FROM node:18-alpine
+FROM node:18-slim
```
Comment (minottic) [inline Dockerfile:1]: I am unsure I understand why `slim` is needed, but not a big deal

File: test/routes.ts
Comment (minottic) [inline test/routes.ts:1]: why this refactor?

File: test/fixtures/returns/panosc.returns.ts
Comment (minottic) [inline test/fixtures/returns/panosc.returns.ts:1]: why were these changes needed?

**Review by minottic — COMMENTED** [review]
some more comments, but it's almost ready I think.

I am unsure I understand why the tests changes were needed

**Comment by minottic** [comment]
I think the push to ghcr is failing because it's happening from a fork

File: Dockerfile (hunk 14 above)
Comment (PatrickFuhrmann-HTWBerlin) [reply Dockerfile:1]: Very good point.  Let me show:

- Failed [apline](https://github.com/PatrickFuhrmann-HTWBerlin/oai-provider-service/actions/runs/10059807904/job/27805867258)
- Succeeded [slim](https://github.com/PatrickFuhrmann-HTWBerlin/oai-provider-service/actions/runs/10059887570/job/27806127477)

File: test/routes.ts
Comment (PatrickFuhrmann-HTWBerlin) [reply test/routes.ts:1]: 1. **Change of the tests loop to individual blocks**. It turned out that the code differences between the four test cases ( 'ListRecords' 'ListIdentifiers''Identify','GetRecord') became so significant that it looked very artificial to stay with the single body and loop over const tests = [ ... ]. Examples like 'if(mock) and if(spy) etc. 
2. **Remove Date Block**. The problem is that we compare output with static content. (Fixtures). As soon as you have dates, this doesn't work anymore. The first date was already removed before. I removed another date.

File: .github/workflows/ci.yml (hunk 13 above)
Comment (PatrickFuhrmann-HTWBerlin) [reply .github/workflows/ci.yml:4]: Sure. I wanted to make sure that the test/deploy works on your side.

File: test/fixtures/returns/panosc.returns.ts
Comment (PatrickFuhrmann-HTWBerlin) [reply test/fixtures/returns/panosc.returns.ts:1]: For some reason, the output of the tests were not exactly identical to the fixtures. Sometimes it was only a CR compared to a blank or capital/small letters. But there were even more significant cases. E.G. with listRecords. In this case the XML namespaces are listed in the <OAI-PMH as well as in the <panosc:panosc_dc clause. However with the new library the output of the  <panosc:panosc_dc is no longer printing all the namespaces but only those which haven't been  displayed yet in the <OAI-PMH clause. In the fixtures they are all listed. Consequently it reports an error.

File: Dockerfile (hunk 14 above)
Comment (bpedersen2) [reply Dockerfile:1]: Looks like alpine does not  have the arm emulation  support needed

Comment (PatrickFuhrmann-HTWBerlin) [reply Dockerfile:1]: That's what I thought, but 'docker manifest inspect node:18-slim' and 'docker manifest inspect node:18-apline' are essentially showing the same supported architectures.

File: .github/workflows/deploy.yml [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+          images: ghcr.io/${{ github.repository }}
+          flavor: latest=true # adds :latest tag to outputs.tags
+          tags: type=sha,format=long,prefix= # adds :<sha> tag to outputs.tags
+
+      - name: Build and push
+        uses: docker/build-push-action@v6
+        with:
+          context: .
+          platforms: linux/amd64,linux/arm64/v8
+          push: true
```
Comment (minottic) [inline .github/workflows/deploy.yml:41]: can we set this to true only on push to master? in this way, we avoid the permission error when coming from a fork

Comment (PatrickFuhrmann-HTWBerlin) [reply .github/workflows/deploy.yml:42]: Yes, I think we can:

- eac0a72 :  made the push dependent on push to master.
- 372bfc7: forgot to change the trigger back to master.

Comment (minottic) [reply .github/workflows/deploy.yml:42]: I meant the the push should be conditional depending on the action trigger. True if push, false if pr

**Review by minottic — APPROVED** [review]
thx!

File: .github/workflows/deploy.yml (hunk 15 above)
Comment (PatrickFuhrmann-HTWBerlin) [reply .github/workflows/deploy.yml]: Sorry I misunderstood. I fixed with [077150a](https://github.com/SciCatProject/oai-provider-service/pull/112/commits/077150ad9cf7dd90d847811fb1cac5c7c675cffa). Seems to work for 'push' on my branch and 'pr' on the scicat master.

Comment (minottic) [reply .github/workflows/deploy.yml]: np! thanks for the changes! I would be fine merging now, do you plan any other change or is it good to go?

Comment (PatrickFuhrmann-HTWBerlin) [reply .github/workflows/deploy.yml]: I'm not planning any other changes in this PR so we are good to go from my side. Thanks a lot for your help.
